roles/upgrade/pre-upgrade/tasks/main.yml

```diff
@@ -1,4 +1,24 @@
 ---
+- name: Get current Kubernetes server version
+  command: "{{ bin_dir }}/kubectl version -o json"
+  register: kubectl_version
+  changed_when: false
+  when: inventory_hostname in groups['kube-master']
+
+- name: Check that current Kubernetes version is supported
+  vars:
+    client_version: "{{ (kubectl_version.stdout | from_json).clientVersion.gitVersion }}"
+    server_version: "{{ (kubectl_version.stdout | from_json).serverVersion.gitVersion }}"
+  assert:
+    that:
+      - client_version is version(kube_version_min_required, '>=')
+      - server_version is version(kube_version_min_required, '>=')
+    msg: "Your version of Kubernetes is not supported for upgrade. Minimum version {{ kube_version_min_required }}"
```

Comment:
wat ? why ?

current kubespray  will update the cluster version 1.16 without problems

Reply:
Kubeadm 1.19 will only update 1.18.0 or later. Since master is 1.19 there is no point in maintaining support for 1.17

Reply:
set in inventory:
```
kube_version: 1.17.12
```
and play upgrade-cluster.yaml

next, set `kube_version` to 1.18.9 and play again